### PR TITLE
[v0.26.0rc][Performance][LoRA] Add A3 single-adapter masked GEMM fast path

### PR DESCRIPTION
**Pull request by Liuchenbing-2026**
> **v0.26.0rc backport** of main PR #13496.
> Targets `releases/v0.26.0rc`. Companion main PR: #13496.

## What this PR does

A3 single-adapter performance optimization. Adds the masked GEMM fast path for `max_loras=1`, non-fully-sharded LoRA on A3:

- Metadata stage precomputes BF16/FP16 adapter mask; graph reuses a fixed-address mask buffer.
- Mask hoisted from the wide output delta to the rank=8 shrink.
- Packed QKV / Gate-Up projection with one contiguous writeback.
- Adapter create/set/reset maintains packed LoRA-A; multiple LoRA-A merged into one shrink GEMM.
- `scale == 1.0` fast path: direct `add_`/`copy_`, drops the no-op scalar multiply.
- Multi-adapter, fully-sharded, and non-qualifying requests keep the original fallback.

## Eligibility

```
device = A3
max_loras = 1
fully_sharded_loras = false
```

## Performance (historical A3)

| Concurrency | Original PR #12295 | Optimized | Δ |
|---:|---:|---:|---:|
| C1 | 29.33 tok/s | 33.27 tok/s | **+13.4%** |
| C4 | 82.15 tok/s | 107.45 tok/s | **+30.8%** |
| C8 | 134.36 tok/s | 173.09 tok/s | **+28.8%** |
| C16 | 199.29 tok/s | 252.66 tok/s | **+26.8%** |

Two A3 environments reproduce within 0.2%-0.3%.

## Scope

Builds on PR 2 (`split/lora-dual-acl-graph`). Does not include A2 enablement (PR 4).

## Validation

```
Ruff check/format: Pass
mypy Python 3.10/3.11/3.12: Pass
Focused pytest: 29 passed
git diff --check: Pass
```

- vLLM version: v0.26.0
- vLLM main: https://github.com/vllm-project/vllm/commit/d02df748bf9efd99022f1a062597dc3cb3808485

Co-Authored-By: wangzhao-11a <73340653+wangzhao-11a@users.noreply.github.com>

### CODE REVIEW

File: vllm_ascend/lora/punica_npu.py [hunk 1]
```diff
@@ -48,6 +49,40 @@ def __init__(self, max_num_batched_tokens: int, max_batches: int, device: torch.
         self.sgmv_expand = sgmv_expand
         self.sgmv_expand_slice = sgmv_expand_slice
         self.sgmv_shrink = sgmv_shrink
+        self._single_lora_slot = (
+            ascend_device_type == AscendDeviceType.A3
```
Comment (voidvelocity) [inline vllm_ascend/lora/punica_npu.py:53]: If this optimized GEMM can be used on A2, add A2 as well.

File: vllm_ascend/lora/punica_npu.py [hunk 2]
```diff
@@ -48,6 +49,40 @@ def __init__(self, max_num_batched_tokens: int, max_batches: int, device: torch.
...
+    ) -> None:
+        super()._update_base_metadata(mapping, lora_index_to_id, max_loras, vocab_size)
+        if self._single_lora_mask is None:
+            return
+
+        token_count = self.indices_len[0]
+        assert token_count is not None
+        token_indices = torch.narrow(self._token_lora_indices, 0, 0, token_count)
+        mask = torch.narrow(self._single_lora_mask, 0, 0, token_count)
+        mask.copy_(token_indices.eq(0).unsqueeze(1))
```
Comment (voidvelocity) [inline vllm_ascend/lora/punica_npu.py]: I'd recommand this way to keep performance and remain readability.

```python
self._single_lora_mask[:token_count].copy_(
    self._token_lora_indices[:token_count].eq(0).unsqueeze(1)
)
```

File: vllm_ascend/lora/punica_npu.py [hunk 3]
```diff
@@ -147,6 +182,10 @@ def _expand_slice_decode(
     def _get_token_lora_indices(self, x: torch.Tensor) -> torch.Tensor:
         return torch.narrow(self._token_lora_indices, 0, 0, x.size(0))
 
+    def _get_single_lora_mask(self, x: torch.Tensor) -> torch.Tensor:
+        assert self._single_lora_mask is not None
+        return torch.narrow(self._single_lora_mask, 0, 0, x.size(0))
```
Comment (voidvelocity) [inline vllm_ascend/lora/punica_npu.py]: I don't think it's necessary to write a new function to get `self._single_lora_mask` since this function only be used once and directly return `self._single_lora_mask`. Try to remove it.

File: vllm_ascend/lora/utils.py [hunk 4]
```diff
@@ -1,19 +1,116 @@
...
+    @classmethod
+    @_not_fully_sharded_can_replace
+    def can_replace_layer(
+        cls,
+        source_layer: nn.Module,
+        lora_config: LoRAConfig,
+        packed_modules_list: list,
+        model_config: PretrainedConfig | None,
+    ) -> bool:
+        return (
```
Comment (voidvelocity) [inline vllm_ascend/lora/utils.py:87]: I think the class that inherits from `_PackedLoRAAWeightsMixin` should only be applied to **single lora slot**, so the `can_replace_layer` should check if only one single lora slot.

File: vllm_ascend/lora/utils.py [hunk 5]
```diff
@@ -1,19 +1,116 @@
...
+    @classmethod
+    @_not_fully_sharded_can_replace
+    def can_replace_layer(
+        cls,
+        source_layer: nn.Module,
+        lora_config: LoRAConfig,
+        packed_modules_list: list,
+        model_config: PretrainedConfig | None,
+    ) -> bool:
+        return type(source_layer) is AscendQKVParallelLinear and len(packed_modules_list) == 3
```
Comment (voidvelocity) [inline vllm_ascend/lora/utils.py]: The same question. I think the class that inherits from `_PackedLoRAAWeightsMixin` should only be applied to **single lora slot**, so the `can_replace_layer` should check if only one single lora slot.

**Review by voidvelocity — COMMENTED** [review]
This PR provides the performance data, I think at least a accurany test should be provided to make sure no precision problem.

**Review by voidvelocity — APPROVED** [review]
This PR add a branch for gemm when only a LoRA, good now.

**Review by kunpengW-code — APPROVED** [review]
